src/meta_schedule/search_strategy/evolutionary_search.cc

```diff
@@ -374,23 +381,25 @@ class EvolutionarySearchNode : public SearchStrategyNode {
   static constexpr const char* _type_key = "meta_schedule.EvolutionarySearch";
   TVM_DECLARE_FINAL_OBJECT_INFO(EvolutionarySearchNode, SearchStrategyNode);
 
-  void InitializeWithTuneContext(const TuneContext& context) final {
-    CHECK(context.defined()) << "TuneContext must be defined!";
-    CHECK(context->num_threads > 0) << "Number of threads has to be larger than 0.";
-    CHECK(context->target.defined()) << "Target must be defined!";
-    this->context_ = context.get();
-    this->rand_state_ = ForkSeed(&context->rand_state);
-    for (const auto& kv : context->mutator_probs) {
-      double mass = kv.second->value;
-      TVM_META_SCHEDULE_CHECK_PROB_RANGE(mass, "mutator_probs");
-    }
+  void InitializeWithTuneContext(const TuneContext& ctx) final {
+    CHECK(ctx->num_threads > 0) << "ValueError: `TuneContext.num_threads` must be > 0";
+    CHECK(ctx->space_generator.defined())
+        << "ValueError: `TuneContext.space_generator` must be defined";
+    CHECK(ctx->space_generator.value()->postprocs.defined())
+        << "ValueError: `TuneContext.space_generator.postprocs` must be defined";
+    CHECK(ctx->space_generator.value()->mutator_probs.defined())
+        << "ValueError: `TuneContext.space_generator.mutator_probs` must be defined";
+    this->ctx_ = ctx.get();
+    this->postprocs_ = ctx->space_generator.value()->postprocs.value();
```

Comment:
I noticed that`sch_rules`, `postprocs` and `mutator_probs` are all moved into space generator. I'm happy to see the change for `sch_rules` as they are only used in space generators; on the other hand, postprocs and mutator_probs are used in search strategy class if not otherwise specified, this creates extra dependecy of search strategy over space generator. Is there any advantage of moving it into space generator compared to leave them inside of TuneContext?

Reply:
After discussing with Bohan, TQ and many others, we found that those three rules depicts the property of a search space (e.g. mutator describes connectivity, postproc is 1-1 transformation from one point to another in the search space), so we consolidated them into one place